Add service level config change subscription API

eliransin opened the pull request:
In order to decouple the service level controller from the systems logic, we introduce an API for subscribing to configuration changes. The timing of the call was determined with resource creation and destruction in mind. An API subscriber can create
resources that will be available from the very start of the service level existence it can also destroy them since the service level
is guarantied not to exist anymore at the time of the call to the deletion notification callback.

Testing:
unit tests - all + a newly added one.
dtests - next-gating (dev mode)

psarna commented on service/qos/service_level_controller.hh:
```diff
@@ -67,6 +68,23 @@ public:
     };
     using service_level_distributed_data_accessor_ptr = ::shared_ptr<service_level_distributed_data_accessor>;
+
+    class subscriber_registry {
```
Why the wrapper? It's more common to keep a vector of subscriber pointers directly, which makes the API simpler and avoid another level of indirection. Example: https://github.com/scylladb/scylla/blob/b503ec36c22856cf9e1d02ff0055767001be6f66/service/endpoint_lifecycle_subscriber.hh#L93

eliransin replied:
I thought that a race might occur, but the more I think about it It probably can't happen since the tasks are executed in an
orderly fashion.
For completeness I will describe this (impossible) race that I had in mind. However, I am going to change the code.
1. A certain object registers for the API.
2. A parallel for each is queuing a call to this object's callback (which is not executed)
3. The object preregisters and destroys itself.
4. Then the queued task from 2 is executed on a dangling pointer.

The thing I didn't consider was that the tasks can't be executed out of order so it can't happen :slightly_smiling_face:

eliransin replied:
removed in the new version

psarna commented on service/qos/service_level_controller.hh:
```diff
@@ -67,6 +68,23 @@ public:
+    class subscriber_registry {
+    private:
+        qos_configuration_change_subscriber* _qccs;
+        friend class service_level_controller;
```
I still think that this whole wrapper class is unnecessary, but if it isn't then I have another question - is this friendship used anywhere? I don't see where.

eliransin replied:
It is used in the parallel for each in the `service_level_controller::notify_*` functions but I am going to remove this so never mind :slightly_smiling_face:

eliransin replied:
removed in the new version

avikivity commented on service/qos/qos_configuration_change_subscriber.hh:
```diff
@@ -0,0 +1,14 @@
+#pragma once
+#include "qos_common.hh"
```
Missing copyright/license blurb.

eliransin replied:
Will add

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -249,7 +274,18 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
+        }
+    }).then([this] () {
+        std::remove_if(_subscribers.begin(), _subscribers.end(), [] (subscriber_registry_entry& e) {
+            return *e == (qos_configuration_change_subscriber*)nullptr;
```
Why this cast?

eliransin replied:
dito

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -230,16 +230,41 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
+        }).then([this, sl_it, slo] () {
+            std::remove_if(_subscribers.begin(), _subscribers.end(), [] (subscriber_registry_entry& e) {
+                return *e == (qos_configuration_change_subscriber*)nullptr;
+            });
```
I don't understand this remove_if. First, std::remove_if() doesn't remove anything, it just rearranges the input. Second, why remove subscribers on service level updates?

eliransin replied:
It is removed in the new version, I tried to reinvent the wheel for protection from deregistration. It was a bad idea...

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -230,16 +230,41 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
     future<> f = make_ready_future();
     if (sl_it != _service_levels_db.end()) {
-        sl_it->second.slo = slo;
+        service_level_options slo_before = sl_it->second.slo;
```
Are these notifications serialized? If not, then a later notification could execute earlier than a previous notification.

Are notifications protected against unsubscribes? We have atomic_vector (that @psarna also mentioned) to assist with this.

eliransin replied:
The serialization is guarantied, this is guarantied even before this change.  The calls to all of the "notify*" functions are serialized.
I have amended the code to use atomic_vector.

eliransin commented:
@psarna / @avikivity you probably forgot :slightly_smiling_face:  I have addresed all comments and waiting for your review.
Tx

avikivity commented:
> @psarna / @avikivity you probably forgot  I have addresed all comments and waiting for your review.
> Tx

If you change something, tell us that you changed it (and what changed, I don't want to reverse-engineer the changes).

eliransin commented:
Ahh sorry @avikivity :
New version:
1. Changed the register / unregister mechanism to work with atomic_vector (as in other places)
2. Added copyright blurbs where it was missing missing

psarna commented on service/qos/service_level_controller.cc:
```diff
@@ -231,15 +231,30 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
+        _subscribers.for_each([name, sl_data] (qos_configuration_change_subscriber* subscriber) {
+            subscriber->on_before_service_level_add(name, sl_data.slo).get();
+        });
+    }).then([this, name, sl_data] () {
```
nitpick: you already created a thread, so you can just as well call `_service_levels_db.emplace(name, sl_data);` inside the thread as well, which saves us some lines and an allocation

psarna replied:
Also - perhaps we should consider what happens if one of the `on_before_service_level_add` callbacks throws - currently, we won't emplace a new service level to `_service_levels_db` - is that expected? Or should we just log the error and still try to emplace the service level? Same question applies to `on_before_service_level_change` below

eliransin replied:
You are right, we should at least try to continue.

psarna commented on service/qos/service_level_controller.cc:
```diff
@@ -231,15 +231,30 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
+            _subscribers.for_each([name, slo_before, slo] (qos_configuration_change_subscriber* subscriber) {
+                subscriber->on_before_service_level_change(name, slo_before, slo).get();
+            });
+        }).then([this, sl_it, slo] () {
```
ditto - let's move contents of this continuation into the thread, since one was already created

eliransin commented:
Changes in the last version:
1. Error out when an exception is thrown from a callback
2. Fold unnecessary then into the already created seastar then

psarna commented on service/qos/service_level_controller.cc:
```diff
@@ -231,15 +231,34 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
+            try {
+                subscriber->on_before_service_level_add(name, sl_data.slo).get();
+            } catch(...) {
+                sl_logger.error("notify_service_level_added: exeption occured in one of the obesrvers callbacks");
```
nitpicks: `s/exeption/exception`, `s/occured/occurred`, `s/obesrvers/observers`

avikivity replied:
Also: log the exception

catch needs a space after it (it's not a function)

psarna commented on service/qos/service_level_controller.cc:
```diff
@@ -231,15 +231,34 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
+                try {
+                    subscriber->on_before_service_level_change(name, slo_before, slo).get();
+                } catch(...) {
+                    sl_logger.error("notify_service_level_updated: exeption occured in one of the obesrvers callbacks");
```
ditto

psarna commented on service/qos/service_level_controller.cc:
```diff
@@ -249,7 +268,15 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
+            try {
+                subscriber->on_after_service_level_remove(name).get();
+            } catch(...) {
+                sl_logger.error("notify_service_level_removed: exeption occured in one of the obesrvers callbacks");
```
ditto

psarna commented on service/qos/qos_configuration_change_subscriber.hh:
```diff
@@ -0,0 +1,56 @@
+ */
+
+/*
+ * Copyright (C) 2015-present ScyllaDB
```
2021

eliransin replied:
will change

psarna commented on test/boost/service_level_controller_test.cc:
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2015-present ScyllaDB
```
2021

psarna reviewed (approved):
Left a couple of typo nitpicks

avikivity commented on service/qos/qos_configuration_change_subscriber.hh:
```diff
@@ -0,0 +1,56 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Why are you licensing this to the ASF?

eliransin replied:
by mistake, copy paste is the devil :blush:

avikivity commented on service/qos/qos_configuration_change_subscriber.hh:
```diff
@@ -0,0 +1,56 @@
+namespace qos {
+    class qos_configuration_change_subscriber {
+    public:
+        /** This callback is going to be called just before the service level is available **/
```
Abstract base classes typically need a virtual destructor.

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -249,7 +268,15 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
     if (sl_it != _service_levels_db.end()) {
         _service_levels_db.erase(sl_it);
```
Is this called under some lock? Otherwise different notifications will get mixed up.

eliransin replied:
Yes, at the top level, there is a serialization semaphore (with one units). The global controller grabs it for every notification it sends and releases it when it is done so every shard can handle only one notification simultaneously.
This semaphore is at: `service_level_controller::global_controller_data::notifications_serializer

eliransin commented:
New in this version:
1. Add a virtual constructor to the virtual observer interface class
2. log the exceptions that happen on observers
3. Typo and coding style fixes
4. Fix licence blurb

avikivity commented on service/qos/service_level_controller.hh:
```diff
@@ -220,11 +222,23 @@ private:
+    future<> unregister_subscriber(qos_configuration_change_subscriber* subscriber);
+
     static sstring default_service_level_name;
 
```
I see you made this public, was this intended?

eliransin replied:
It was intended but maybe my intentions are wrong?
We need a subscriber or someone on his behalf will call this API, so it's either that or friendship of `qos_configuration_change_subscriber`

avikivity replied:
I meant `default_service_level_name`. Is it now part of the API?

avikivity replied:
Wait, it was public before.

avikivity commented on test/boost/service_level_controller_test.cc:
```diff
@@ -0,0 +1,130 @@
+#include "service/qos/qos_configuration_change_subscriber.hh"
+#include "auth/service.hh"
+
+// helper type for the visitor #4
```
What is `#4`?

You can use the overloaded_functor helper we have.

eliransin replied:
> What is `#4`?

A copy paste error :relaxed: 

> You can use the overloaded_functor helper we have.


I will do that

eliransin replied:
done
